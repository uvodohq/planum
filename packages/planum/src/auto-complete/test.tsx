import {
  autoUpdate,
  size,
  useDismiss,
  useFloating,
  useId,
  useInteractions,
  useListNavigation,
  useRole,
} from '@floating-ui/react-dom-interactions'
import React, { forwardRef, useLayoutEffect, useRef, useState } from 'react'

interface ItemProps {
  children: React.ReactNode
  active: boolean
}

const Item = forwardRef<
  HTMLDivElement,
  ItemProps & React.HTMLProps<HTMLDivElement>
>(({ children, active, ...rest }, ref) => {
  const id = useId()
  return (
    <div
      ref={ref}
      role="option"
      id={id}
      aria-selected={active}
      {...rest}
      style={{
        background: active ? 'lightblue' : 'none',
        padding: 4,
        cursor: 'default',
        ...rest.style,
      }}>
      {children}
    </div>
  )
})
Item.displayName = 'Item'

export function AutoCompleteTest({ data }) {
  const [open, setOpen] = useState(false)
  const [inputValue, setInputValue] = useState('')
  const [activeIndex, setActiveIndex] = useState<number | null>(null)

  const listRef = useRef<Array<HTMLElement | null>>([])

  const { x, y, reference, floating, strategy, context, refs } =
    useFloating<HTMLInputElement>({
      whileElementsMounted: autoUpdate,
      open,
      onOpenChange: setOpen,
      middleware: [
        size({
          apply({ rects, elements }) {
            Object.assign(elements.floating.style, {
              width: `${rects.reference.width}px`,
              maxHeight: `${200}px`,
            })
          },
          padding: 10,
        }),
      ],
    })

  useLayoutEffect(() => {
    // IMPORTANT: When the floating element first opens, this runs when the
    // styles have **not yet** been applied to the element. This can cause an
    // infinite loop as `size` has not yet limited the maxHeight, so the whole
    // page tries to scroll. We must wrap it in rAF.
    requestAnimationFrame(() => {
      if (activeIndex != null) {
        listRef.current[activeIndex]?.scrollIntoView({ block: 'nearest' })
      }
    })
  }, [activeIndex])

  const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions(
    [
      useRole(context, { role: 'listbox' }),
      useDismiss(context),
      useListNavigation(context, {
        listRef,
        activeIndex,
        onNavigate: setActiveIndex,
        virtual: true,
        loop: true,
        focusItemOnHover: false,
      }),
    ],
  )

  function onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const value = event.target.value
    setInputValue(value)

    if (value) {
      setOpen(true)
      setActiveIndex(0)
    } else {
      setOpen(false)
    }
  }

  const items = data.filter((item) =>
    item.toLowerCase().startsWith(inputValue.toLowerCase()),
  )

  return (
    <>
      <input
        {...getReferenceProps({
          ref: reference,
          onChange,
          value: inputValue,
          placeholder: 'Enter fruit',
          'aria-autocomplete': 'list',
          onKeyDown(event) {
            if (
              event.key === 'Enter' &&
              activeIndex != null &&
              items[activeIndex]
            ) {
              setInputValue(items[activeIndex])
              setActiveIndex(null)
              setOpen(false)
            }

            if (event.key === 'Tab') {
              setOpen(false)
            }
          },
        })}
      />
      {open && (
        <div
          {...getFloatingProps({
            ref: floating,
            style: {
              position: strategy,
              left: x ?? 0,
              top: y ?? 0,
              background: '#eee',
              color: 'black',
              overflowY: 'auto',
            },
          })}>
          {items.map((item, index) => (
            // eslint-disable-next-line react/jsx-key
            <Item
              {...getItemProps({
                key: item,
                ref(node) {
                  listRef.current[index] = node
                },
                onClick() {
                  setInputValue(item)
                  setOpen(false)
                  refs.reference.current?.focus()
                },
              })}
              active={activeIndex === index}>
              {item}
            </Item>
          ))}
        </div>
      )}
    </>
  )
}
