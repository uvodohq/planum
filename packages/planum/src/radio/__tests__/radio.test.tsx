// import { vi } from 'vitest'

import { render } from '../../test'
import { Radio, RadioGroup } from '..'

// vi.mock('@react-aria/utils', async () => {
//   console.log('1')
//   const actual: any = await vi.importActual('@react-aria/utils')
//   console.log('2')

//   console.log(actual)
//   return {
//     ...actual,
//     useId: () => 'test-ID',
//     useSlotId: () => 'test-slot-ID',
//     useDrag1D: () => 'test-drag-ID',
//   }
// })

describe('Radio', async () => {
  it.todo('should render 3 radios - cant mock ID', () => {
    const { container } = render(
      <RadioGroup aria-label="label">
        <Radio value="Regular">Regular</Radio>
        <Radio value="Selected">Selected</Radio>
      </RadioGroup>,
    )
    const label = container.querySelectorAll('label')
    expect(label.length).toBe(2)
    expect(container).toMatchSnapshot()
  })
})
