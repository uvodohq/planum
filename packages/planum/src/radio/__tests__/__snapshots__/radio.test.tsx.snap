// Vitest Snapshot v1

exports[`Radio > should render 3 radios 1`] = `
<div>
  <div
    aria-label="label"
    aria-orientation="vertical"
    class="c-PJLV"
    id="react-aria224783748-1"
    role="radiogroup"
  >
    <label
      class="c-iPYhUp c-kLRYwF c-TiAfA"
    >
      <div
        style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
      >
        <input
          name="react-aria224783748-5"
          tabindex="0"
          type="radio"
          value="Regular"
        />
      </div>
      <svg
        aria-hidden="true"
        class="c-fApySl"
      >
        <circle
          cx="10"
          cy="10"
          fill="currentColor"
          r="4"
          stroke="none"
          stroke-width="2"
        />
      </svg>
      <p
        class="c-iPYhUp c-kLRYwF c-iPYhUp-ikmBbln-css"
      >
        Regular
      </p>
    </label>
    <label
      class="c-iPYhUp c-kLRYwF c-TiAfA"
    >
      <div
        style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
      >
        <input
          name="react-aria224783748-5"
          tabindex="0"
          type="radio"
          value="Selected"
        />
      </div>
      <svg
        aria-hidden="true"
        class="c-fApySl"
      >
        <circle
          cx="10"
          cy="10"
          fill="currentColor"
          r="4"
          stroke="none"
          stroke-width="2"
        />
      </svg>
      <p
        class="c-iPYhUp c-kLRYwF c-iPYhUp-ikmBbln-css"
      >
        Selected
      </p>
    </label>
  </div>
</div>
`;
